component-base: make LoggingConfiguration a single-version API

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

Making the LoggingConfiguration part of the versioned component-base/config API
had the theoretic advantage that components could have offered different
configuration APIs with experimental features limited to alpha versions (for
example, sanitization offered only in a v1alpha1.KubeletConfiguration). Some
components could have decided to only use stable logging options.

In practice, this wasn't done. Furthermore, we don't *want* different components
to make different choices regarding which logging features they offer to
users. It should always be the same everywhere, for the sake of consistency.

This can be achieved by dropping the distinction between
internal and external LoggingConfiguration types. The Go API becomes simpler, too.
Different stability levels of individual fields have to be covered by documentation and
feature gates.

Advantages:

- everything related to logging is under component-base/logs;
  previously this was scattered across different packages and
  different files under "logs" (why some code was in logs/config.go
  vs. logs/options.go vs. logs/logs.go always confused me again
  and again when coming back to the code):

  - long-term config and command line API are clearly separated
    into the "api" package underneath that

  - logs/logs.go itself only deals with legacy global flags and
    logging configuration

- removal of separate Go APIs like logs.BindLoggingFlags and
  logs.Options

- LogRegistry becomes an implementation detail, with less code
  and less exported functionality (only registration needs to
  be exported, querying is internal)

#### Special notes for your reviewer:

This is an alternative to PR https://github.com/kubernetes/kubernetes/pull/105448

Feature checking is implemented as in other components by accessing the global default feature gate. For that, that instance has to be moved to component-base to avoid circular dependencies. The alternative is to pass in a gate instance for checking, but that means changing the API without providing any tangible advantage. That approach has been implemented for comparison in https://github.com/pohly/kubernetes/commits/log-configuration-feature-gates-optional and (for some other PR) in https://github.com/pohly/kubernetes/commits/log-contextual-feature-gate-optional. See https://kubernetes.slack.com/archives/C0EG7JC6T/p1644329116043899?thread_ts=1643908087.471269&cid=C0EG7JC6T



#### Does this PR introduce a user-facing change?
```release-note
The Go API for logging configuration in k8s.io/component-base was moved to k8s.io/component-base/logs/api/v1. The configuration file format and command line flags are the same as before.
```

pohly committed:
logs: make LoggingConfiguration an unversioned API

Making the LoggingConfiguration part of the versioned component-base/config API
had the theoretic advantage that components could have offered different
configuration APIs with experimental features limited to alpha versions (for
example, sanitization offered only in a v1alpha1.KubeletConfiguration). Some
components could have decided to only use stable logging options.

In practice, this wasn't done. Furthermore, we don't want different components
to make different choices regarding which logging features they offer to
users. It should always be the same everywhere, for the sake of consistency.

This can be achieved with a saner Go API by dropping the distinction between
internal and external LoggingConfiguration types. Different stability levels of
indidividual fields have to be covered by documentation (done) and potentially
feature gates (not currently done).

Advantages:

- everything related to logging is under component-base/logs;
  previously this was scattered across different packages and
  different files under "logs" (why some code was in logs/config.go
  vs. logs/options.go vs. logs/logs.go always confused me again
  and again when coming back to the code):

  - long-term config and command line API are clearly separated
    into the "api" package underneath that

  - logs/logs.go itself only deals with legacy global flags and
    logging configuration

- removal of separate Go APIs like logs.BindLoggingFlags and
  logs.Options

- LogRegistry becomes an implementation detail, with less code
  and less exported functionality (only registration needs to
  be exported, querying is internal)

pohly commented:
I think this worked as intended.

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/106123/pull-kubernetes-e2e-windows-gce/1455925818994724864/artifacts/e2e-7d7d9eefc0-95646-windows-node-group-f4fn/serial-1.log shows that kubelet and kube-proxy get started via kube-runner.

dashpole commented:
/assign @serathius 
/triage accepted

serathius commented:
I think API review should give us idea if this is a good direction
/unassign
/label api-review

ehashman commented:
/assign

pohly commented:
/retest

pohly commented:
/retest

dims commented:
This is in api review! do we want to target this for 1.24?

pohly commented:
Yes, let's target 1.24.

neolit123 commented:
We had similar problems in kubeadm.

The kubeadm component config used to be v1alpha3 and embedded the kubeproxyconfiguration v1alpha1. We wanted to graduate the kubeadm config to beta, but the embedding meant there are alpha portions in it and it did not feel right to proceed like that. So what we did is we started accepting multiple APIs (GVK) in a multi-doc. This allowed us to decouple the maturity of the APIs. It feels like the best option.

In another instance we force graduated a common structure that was embedded in multiple groups to a v1 object in a new group. This meant we can reuse it in the groups that need it. But had questions wether it's actualy stable and what happens if we want to change it in a breaking way. Do we keep the v1 forever, if it changes to v2 do we bump the parent APIs to v2 as well? Etc.. These complications come from embedding.

For this logging configuration having it unversioned seems fine to me, but at minimum it must have a README explaining what it is, where it is used, why it is unversioned, and that ideally users should not import it. But embedding is still problematic. If the kubeletconfiguration one day graduates to v1 and continues to embed this, it means that removing a field from the logging configuration is effectively breaking a v1 struct.

neolit123 commented:
> Huh? If users don't import the API package, how do they use it? I probably misunderstand this part... 😅

AFAIK curtently it is used only by and meant for k8s core components. And maybe users should not import it standalone to use it in their custom tools. Or maybe the intention is different here.

Either way, if it becomes unversioned, its version effectively becomes that of the tag in component-base, so at least external users (even if discouraged) can track changes that way.

neolit123 commented:
BTW, this is a breaking change to users who bake e.g a kubeletconfiguration
from go structs and the PR would need "action required".

neolit123 commented:
> I think it is different. The LoggingConfiguration struct can also be used by out-of-tree components to define and implement a) their command line options and/or b) a configuration file.

i guess they will have the git tags, so there will be _some_ versioning.

> My impression was that staging repos provide no API guarantees whatsoever and developers need to react to build errors if and when they occur. An "action required" would make more sense to me if the API change didn't lead to build errors. But here it does, so there's no risk that someone misses some crucial step while updating

at least the kubelet, kube-proxy configuration API repositories are all staged and they do have API guarantees.

> It's an API change and the "action required" would be "change your imports from component-base/config/v1alpha1 to component-base/logs/api". But do we really announce such source code changes to Kubernetes users in the release notes?

i'd say yes, but deferring to owners.

the number of the following users might be 0, but this change requires an action for users that embed the kubelet configuration structures and assume the old `componentbaseconfigv1alpha1.LoggingConfiguration` e.g. `myKubletConfig.Logging = myLogging`, where `myLogging` is not the same import and go semantics will prevent the assignment.
https://github.com/kubernetes/kubernetes/pull/105797/files#diff-e7cf3d9bacdf8ddf0db56a9a399993d2fc62897122206ecef17aee62c2e69d8fL971-R971

pohly commented:
> at least the kubelet, kube-proxy configuration API repositories are all staged and they do have API guarantees.

That's different because the versioning of the API is done via the directory name (v1beta1, v1). I was thinking of Go module versioning: that is 0.x for all staging repos, i.e. there isn't a stable API for the code in those repos.

pohly commented:
The discussion about how to handle API structs continued in SIG Architecture and in https://kubernetes.slack.com/archives/CJUQN3E4T/p1642104440006600?thread_ts=1639068949.069800&cid=CJUQN3E4T

The conclusion:
- a single struct is okay
- it should be in a "api/v1" package to indicate that API versioning is done - in a way this is just cosmetic because there won't be any v1alpha1 when adding experimental fields, and hopefully never a v2, but it sends the right message
- we want LoggingBetaOptions and LoggingAlphaOptions as feature gates, similar to CPUManagerPolicyBetaOptions and CPUManagerPolicyAlphaOptions
- there has to be a unit test that ensures compatibility with JSON or YAML files written for older releases

I have switch to "api/v1" and implemented the unit test. I'll do the feature gates next.

While doing so I found a bug in the FlushFrequency config support that I added in 1.23. I fixed those and currently the fixes are in the first two commits of this PR, but probably I should put them into a separate PR?

It probably won't reduce the number of required reviewers, but it would be cleaner and easier to backport.

pohly committed:
logs: add alpha+beta feature gates

It is useful to have the ability to control whether alpha or beta features are
enabled. We can group features under LoggingAlphaOptions and LoggingBetaOptions
because the configuration is designed so that each feature individually must be
enabled via its own option.

Currently, the JSON format itself is beta (graduated in 1.23) but additional
options for it were only added in 1.23 and thus are still alpha:

  $ go run ./staging/src/k8s.io/component-base/logs/example/cmd/logger.go --logging-format=json --log-json-split-stream --log-json-info-buffer-size 1M --feature-gates LoggingBetaOptions=false
  [format: Forbidden: Log format json is BETA and disabled, see LoggingBetaOptions feature, options.json.splitStream: Forbidden: Feature LoggingAlphaOptions is disabled, options.json.infoBufferSize: Forbidden: Feature LoggingAlphaOptions is disabled]

  $ go run ./staging/src/k8s.io/component-base/logs/example/cmd/logger.go --logging-format=json --log-json-split-stream --log-json-info-buffer-size 1M
  [options.json.splitStream: Forbidden: Feature LoggingAlphaOptions is disabled, options.json.infoBufferSize: Forbidden: Feature LoggingAlphaOptions is disabled]

This is the same approach that was taken for CPUManagerPolicyAlphaOptions and
CPUManagerPolicyBetaOptions.

In order to test this without modifying the global feature gate in a test file,
ValidateKubeletConfiguration must take a feature gate as argument.

liggitt commented:
> While doing so I found a bug in the FlushFrequency config support that I added in 1.23. I fixed those and currently the fixes are in the first two commits of this PR, but probably I should put them into a separate PR?

yeah, split to https://github.com/kubernetes/kubernetes/pull/107618 sgtm

is https://github.com/kubernetes/kubernetes/pull/105448 superceded by this now?

ehashman commented:
> While doing so I found a bug in the FlushFrequency config support that I added in 1.23. I fixed those and currently the fixes are in the first two commits of this PR, but probably I should put them into a separate PR?

Yes, please split that. I see there is already a PR here which I commented on: https://github.com/kubernetes/kubernetes/pull/107618

pohly commented:
> is #105448 superceded by this now?

I think this PR makes #105448 obsolete because instead of promoting to v1beta1, we probably should move all structs into their own api/v1 package, similar to this PR here.

But there are several review comments in it #105448 which will be relevant when working on those other component-base structs. I suppose we can close it, those comments aren't going away.

pohly commented:
I have rebased on top of the current https://github.com/kubernetes/kubernetes/pull/107618, which should get merged first.

On top of this PR here I have implemented the proposal from Jordan about enhancing the config file support for FlushFrequency (a new field that accepts a string in the config): https://github.com/pohly/kubernetes/commits/log-flush-frequency-new-field

That's intentionally not a PR yet because I don't want to distract from reviewing the other PRs.

pohly commented:
/retest

kerthcet commented:
/retest

acumino commented on staging/src/k8s.io/component-base/config/v1alpha1/defaults.go:
```diff
@@ -19,7 +19,6 @@ package v1alpha1
 import (
 	"time"
 
-	"k8s.io/apimachinery/pkg/api/resource"
```
Will TODO(#80289) be addressed in this PR like it was being addressed in #105448?

pohly replied:
No, this PR is just about LoggingConfiguration. Once it has gone through API review and got merged, similar changes can and probably should be done for the other structs that are currently in component-base/config.

pohly commented:
/retest

pull-kubernetes-integration still flaky?

liggitt commented:
> pull-kubernetes-integration still flaky?

Yes. #103512

pohly commented on staging/src/k8s.io/component-base/featuregate/feature_gate.go:
```diff
@@ -60,6 +60,16 @@ var (
+	// DefaultMutableFeatureGate is a mutable version of DefaultFeatureGate.
+	// Only top-level commands/options setup should make use of this.
+	// Tests that need to modify feature gates for the duration of their test should use:
+	//   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
```
This probably should be the new import path.

`hack/verify-test-featuregates.sh` also needs to be updated. Right now it complains:
```
Test files may not access mutable global feature gates directly:
./staging/src/k8s.io/component-base/logs/json/register/validate_test.go:64:			defer featuregatetesting.SetFeatureGateDuringTest(t, featuregate.DefaultMutableFeatureGate, v1.LoggingAlphaOptions, test.alphaEnabled)()
...
Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
```

pohly replied:
The script was fine, the code shouldn't have used DefaultMutableFeatureGate. Fixed.

pohly commented:
/retest

ehashman commented:
n.b. I think this is an improvement over the status quo but because this PR is so large I am slowly reviewing it, a little at a time. I will keep at it, I haven't seen anything obviously blocking.

ehashman commented:
/remove-priority backlog
/priority important-longterm
/milestone v1.24

would like this to land this release if possible...

ehashman commented on staging/src/k8s.io/component-base/logs/api/v1/registry.go:
```diff
@@ -0,0 +1,98 @@
+/*
```
Diff with removed corresponding file:

```diff
ehashman@red-dot:~/src/k8s$ git diff HEAD^ HEAD -M -- staging/src/k8s.io/component-base/logs/registry/registry.go staging/src/k8s.io/component-base/logs/api/v1/registry.go
diff --git a/staging/src/k8s.io/component-base/logs/api/v1/registry.go b/staging/src/k8s.io/component-base/logs/api/v1/registry.go
index 31a8995c0b5..7871b1389b9 100644
--- a/staging/src/k8s.io/component-base/logs/api/v1/registry.go
+++ b/staging/src/k8s.io/component-base/logs/api/v1/registry.go
@@ -21,6 +21,8 @@ import (
        "sort"
 
        "github.com/go-logr/logr"
+
+       "k8s.io/component-base/featuregate"
 )
 
 var logRegistry = newLogFormatRegistry()
@@ -34,6 +36,10 @@ type logFormatRegistry struct {
 // LogFormatFactory provides support for a certain additional,
 // non-default log format.
 type LogFormatFactory interface {
+       // Feature returns the name of the feature that controls
+       // support for the log format, if there is one.
+       Feature() featuregate.Feature
+
        // Create returns a logger with the requested configuration.
        // Returning a flush function for the logger is optional.
        // If provided, the caller must ensure that it is called
```

pohly replied:
This enables feature checking for individual logging formats.

pohly replied:
While rebasing onto https://github.com/kubernetes/kubernetes/pull/108995 I tried something else: I added one commit which moves four files unmodified into `api/v1`, then made the necessary changes. This makes the "[component-base: make LoggingConfiguration a single-version API](https://github.com/kubernetes/kubernetes/pull/105797/commits/0735a3fc49a2f1a6e6dc730ad297c35fde23ff8a#)" commit a bit more readable.

The downside is that bisecting can fail one the one commit that doesn't build.

I don't have a strong preference either way and can also squash before merging.

@ehashman : please let me know what you prefer.

liggitt replied:
I have a strong preference for reviewable commits over bisect purity... :)

ehashman commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -0,0 +1,284 @@
+/*
```
Diff with removed corresponding file:

```diff
ehashman@red-dot:~/src/k8s$ git diff HEAD^ HEAD -M -- staging/src/k8s.io/component-base/logs/options.go staging/src/k8s.io/component-base/logs/api/v1/options.godiff --git a/staging/src/k8s.io/component-base/logs/api/v1/options.go b/staging/src/k8s.io/component-base/logs/api/v1/options.go
index 8a5a0568562..b64298ad785 100644
--- a/staging/src/k8s.io/component-base/logs/api/v1/options.go
+++ b/staging/src/k8s.io/component-base/logs/api/v1/options.go
@@ -33,6 +33,7 @@ import (
        "k8s.io/apimachinery/pkg/util/validation/field"
        "k8s.io/apimachinery/pkg/util/wait"
        cliflag "k8s.io/component-base/cli/flag"
+       "k8s.io/component-base/featuregate"
 )
 
 var (
@@ -104,8 +105,19 @@ func (c *LoggingConfiguration) ValidateAsField(fldPath *field.Path) field.ErrorL
                        }
                }
        }
-       if _, err := logRegistry.get(c.Format); err != nil {
+       factory, err := logRegistry.get(c.Format)
+       if err != nil {
                errs = append(errs, field.Invalid(fldPath.Child("format"), c.Format, "Unsupported log format"))
+       } else if factory != nil {
+               feature := factory.Feature()
+               if feature != "" && !featuregate.DefaultFeatureGate.Enabled(feature) {
+                       // Look up alpha/beta from actual feature. Should exist, but the API doesn't guarantee it.
+                       status := "experimental"
+                       if featureSpec, ok := featuregate.DefaultMutableFeatureGate.GetAll()[feature]; ok {
+                               status = string(featureSpec.PreRelease)
+                       }
+                       errs = append(errs, field.Forbidden(fldPath.Child("format"), fmt.Sprintf("Log format %s is %s and disabled, see %s feature", c.Format, status, feature)))
+               }
        }
 
        // The type in our struct is uint32, but klog only accepts positive int32.
@@ -128,7 +140,26 @@ func (c *LoggingConfiguration) ValidateAsField(fldPath *field.Path) field.ErrorL
                }
        }
 
-       // Currently nothing to validate for c.Options.
+       errs = append(errs, c.validateFormatOptions(fldPath.Child("options"))...)
+       return errs
+}
+
+func (c *LoggingConfiguration) validateFormatOptions(fldPath *field.Path) field.ErrorList {
+       errs := field.ErrorList{}
+       errs = append(errs, c.validateJSONOptions(fldPath.Child("json"))...)
+       return errs
+}
+
+func (c *LoggingConfiguration) validateJSONOptions(fldPath *field.Path) field.ErrorList {
+       errs := field.ErrorList{}
+       if !featuregate.DefaultFeatureGate.Enabled(LoggingAlphaOptions) {
+               if c.Options.JSON.SplitStream {
+                       errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", LoggingAlphaOptions)))
+               }
+               if c.Options.JSON.InfoBufferSize.Value() != 0 {
+                       errs = append(errs, field.Forbidden(fldPath.Child("infoBufferSize"), fmt.Sprintf("Feature %s is disabled", LoggingAlphaOptions)))
+               }
+       }
        return errs
 }
 
@@ -150,8 +181,8 @@ func (c *LoggingConfiguration) AddFlags(fs *pflag.FlagSet) {
        // JSON options. We only register them if "json" is a valid format. The
        // config file API however always has them.
        if _, err := logRegistry.get("json"); err == nil {
-               fs.BoolVar(&c.Options.JSON.SplitStream, "log-json-split-stream", false, "[Experimental] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout.")
-               fs.Var(&c.Options.JSON.InfoBufferSize, "log-json-info-buffer-size", "[Experimental] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi).")
+               fs.BoolVar(&c.Options.JSON.SplitStream, "log-json-split-stream", false, "[Alpha] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout.")
+               fs.Var(&c.Options.JSON.InfoBufferSize, "log-json-info-buffer-size", "[Alpha] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi).")
        }
 }
```

pohly replied:
The log format feature check is in the first hunk and checking of the other features is in the second.

Instead of the vague "experimental", the two JSON output options now follow the usual graduation (alpha -> beta -> stable).

ehashman commented on staging/src/k8s.io/component-base/logs/api/v1/kube_features.go:
```diff
@@ -0,0 +1,59 @@
+
+var features = map[featuregate.Feature]featuregate.FeatureSpec{
+	LoggingAlphaOptions: {Default: false, PreRelease: featuregate.Alpha},
+	LoggingBetaOptions:  {Default: true, PreRelease: featuregate.Beta},
```
I'm not sure if I like this... this implies that we will have perma alpha and beta feature gates and it will be difficult to track which logging features are behind which gates, and which ones have graduated.

Is it possible to model this after how we normally do feature gates, where each feature has its own graduation cycle?

pohly replied:
This was discussed before in the context of CPUManagerPolicyAlphaOptions and CPUManagerPolicyBetaOptions, see https://groups.google.com/g/kubernetes-sig-architecture/c/Nxsc7pfe5rw

The key arguments why combining multiple features is better in cases like this are:
- Each of the features gets enabled via a flag or configuration option, not via `--feature-gates`, i.e. fine-grained control over features is still possible.
- It avoids overwhelming the user with too many features listed for `--feature-gates`.
- It avoids breaking user's command line invocations when a feature graduates: `kube-scheduler --feature-gate LogJSONSplitStream=true --log-json-split-stream` would become invalid at some point, but `kube-scheduler --feature-gate LoggingAlphaOptions=true --log-json-split-stream` remains valid.

Regarding "it will be difficult to track which logging features are behind which gates": that is addressed by the alpha/beta comments in LoggingConfiguration and the command line flag descriptions. The feature name could be added to LoggingConfiguration if we want to be really precise. For the command line it might become too long.

ehashman reviewed (commented):
I think I'm about halfway through (21/45 files) so thought I'd hit submit now

pohly commented on staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go:
```diff
@@ -22,12 +25,12 @@ import (
 	// Tests that need to modify feature gates for the duration of their test should use:
 	//   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
-	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
+	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.DefaultMutableFeatureGate
```
The usual pattern for a feature check is:
- define a feature somewhere
- automatically add it to k8s.io/apiserver/pkg/util/feature.DefaultFeatureGate in the file that defines the feature
- check state of the feature in DefaultFeatureGate

component-base currently can't do that because it has no access to k8s.io/apiserver/pkg/util/feature (circular dependency problem). The first commit on this PR solves that by moving the the FeatureGate instance into component-base/featuregate.

This particular change here ensures that users of k8s.io/apiserver/pkg/util/feature continue to work as before.

I've tried some alternatives for this (see links under "Special notes for your reviewer") but IMHO they are worse.

liggitt replied:
I expected to pass in the instance, not move the global instance to component-base

pohly replied:
That is how I implemented it in https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab

It makes the API more complicated and I still don't see the advantage of doing it that way. Why should in-tree packages and other staging packages like k8s.io/cloud-provider be allowed to check a global instance, but not component-base? That seems to derive simply from how the code is currently organized, not because it's best for developers or users.

liggitt replied:
because component-base is not about a _particular_ component, and putting the global feature gate there will mean registering feature flags for particular components into what is supposed to be a component-neutral package

pohly replied:
`k8s.io/apiserver/pkg/util` is also not about any particular component. It just happens to be shared by them.

I'd like to point out one thing: I am not suggesting to move the feature registration into component-base. Someone who consumes component-base/featuregates and nothing else gets an empty component-base/featuregates.DefaultFeatureGate instance.

Features get added to that only when a component also imports the code that defines and expects a certain feature. For component-base/logs that would be the new LoggingAlpha/BetaOptions features and nothing else.

pohly replied:
If a consumer of component-base/featuregates wants an empty FeatureGate, they can construct one, just as they do now. They are not forced to use the default one.

pohly replied:
To recap the SIG Architecture discussion that we just had: using a global FeatureGate instance is seen as a problematic code pattern that shouldn't be repeated in component-base if it can be avoided. So it's not about the location of that DefaultFeatureGate instance, it is about what the desired API of component-base/logs and component-base/cli (because sometimes it sits between main() and component-base/logs) should be. That makes sense to me.

I'll revive my alternative branch (a bit stale now because only rebased this one here) and swap the code. Then we can look at that alternative API and decide whether that's better.

pohly replied:
Cherry-picking https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab worked almost without conflicts. I've pushed a new branch with the initial commit removed and the last one replaced.

The feature gate implementation is now more complex. Previous code with access to a global instance:
```
$ git log -p -n1 log-configuration-feature-gates-2022-03-24  | diffstat
 api/v1/kube_features.go        |   59 +++++++++++++++++++++++++++++++
 api/v1/options.go              |   39 ++++++++++++++++++---
 api/v1/options_test.go         |   10 +++--
 api/v1/registry.go             |    6 +++
 api/v1/types.go                |   11 +++--
 api/v1/validate_test.go        |   43 ++++++++++++++++++++++-
 example/cmd/logger.go          |    2 +
 json/json.go                   |    5 ++
 json/register/validate_test.go |   76 +++++++++++++++++++++++++++++++++++++++++
 9 files changed, 237 insertions(+), 14 deletions(-)
```

Now:
```
$ git log -p -n1 log-configuration-feature-gates-2022-03-24-no-default | diffstat
 cmd/kube-apiserver/app/server.go                                      |    2 
 cmd/kube-controller-manager/app/controllermanager.go                  |    2 
 cmd/kube-scheduler/app/server.go                                      |    3 
 cmd/kubelet/app/server.go                                             |    2 
 pkg/features/kube_features.go                                         |    2 
 pkg/kubelet/apis/config/validation/validation.go                      |    2 
 staging/src/k8s.io/component-base/featuregate/feature_gate.go         |   14 ++
 staging/src/k8s.io/component-base/logs/api/v1/features.go             |   57 ++++++++
 staging/src/k8s.io/component-base/logs/api/v1/features_test.go        |   47 +++++++
 staging/src/k8s.io/component-base/logs/api/v1/options.go              |   51 ++++++-
 staging/src/k8s.io/component-base/logs/api/v1/options_test.go         |   12 +
 staging/src/k8s.io/component-base/logs/api/v1/registry.go             |    7 +
 staging/src/k8s.io/component-base/logs/api/v1/types.go                |   11 +
 staging/src/k8s.io/component-base/logs/api/v1/validate_test.go        |   35 +++++
 staging/src/k8s.io/component-base/logs/example/cmd/logger.go          |    9 +
 staging/src/k8s.io/component-base/logs/json/json.go                   |    5 
 staging/src/k8s.io/component-base/logs/json/register/register_test.go |   65 +++++++++-
 17 files changed, 296 insertions(+), 30 deletions(-)
```

Note that this is with a more forgiving implementation that doesn't panic when the passed in FeatureGate doesn't have the feature. That could be changed to panic because all callers should do the necessary registration, but it seems unnecessary to impose registration as a hard requirement on users of the package when there is a reasonable fallback.

The API change in this PR isn't too bad because not many places use LoggingConfiguration.ValidateAndApply.

Where it gets worse is in the implementation of the contextual logging feature gate, because there component-base/cli.Run calls component-base/logs.InitLogs and that is where the new "ContextualLogging" feature gate gets checked: https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab

@liggitt: please have a look at that commit and the last commit of this PR and let me know whether this is what you prefer or whether you see a simpler way of doing this.

pohly replied:
Speaking of contextual logging, getting that feature completed is more important than this PR here. Can we continue the discussion around feature gates to the PR that I just created for that? See https://github.com/kubernetes/kubernetes/pull/108995.

pohly commented:
Merging of https://github.com/kubernetes/kubernetes/pull/108725 caused some conflicts. I have resolved those, but had to rebase.

pohly commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -81,9 +81,13 @@ func epochMillisTimeEncoder(_ time.Time, enc zapcore.PrimitiveArrayEncoder) {
 
-func (f Factory) Create(options config.FormatOptions) (logr.Logger, func()) {
+func (f Factory) Feature() featuregate.Feature {
+	return logsapi.LoggingBetaOptions
```
I should point out that this will break current users of invocations like `kube-scheduler --logging-format=json`: they need to add `--feature-gate LoggingBetaOptions=true`.

Perhaps it would be better to be a bit less precise and not mark JSON output as beta here?

pohly replied:
Scratch that. The beta feature is enabled by default and the alpha features are something where breaking changes are okay.

dims commented:
/retest

@ehashman if you LGTM, we can go find approvers!

pohly commented:
The feature gate part of this PR is conflicting with a similar change in https://github.com/kubernetes/kubernetes/pull/108995, which now needs to go in first. I had hoped to avoid that conflict by merging this here first, but the timing didn't work out.

liggitt commented:
Swept https://github.com/kubernetes/kubernetes/pull/108995, which looks way more contained for 1.24

Moving this piece to 1.25. With the clarity in https://github.com/kubernetes/kubernetes/pull/108995, I think this will be a lot simpler once that merges

/milestone v1.25

pohly committed:
kubelet: initialize logging even sooner, return error

After the removal of the dynamic kubelet configuration feature it became
possible to initialize logging directly after configuration parsing. The
advantage is that logs emitted by
kubeletconfigvalidation.ValidateKubeletConfiguration and
`klog.InfoS("unsupported configuration ...` already use the intended log
output.

After the code was originally added, Run was replaced by RunE. Taking advantage
of that and returning an error is cleaner.

pohly committed:
logs: simplify flush daemon startup

StartFlushDaemon can be called more than once. We can take advantage of that an
eliminate the tracking of where we need to start it.

pohly committed:
logs: move code in preparation for v1 api

kerthcet commented:
`There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors (0/39 nodes are available: 1 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate, 1 node(s) were unschedulable, 3 node(s) had taint {dedicated: greenhouse}, that the pod didn't tolerate, 34 Insufficient cpu, 5 Insufficient memory.)`
Seems resource exhausted. 
/retest

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -213,9 +215,9 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
-				return fmt.Errorf("failed to initialize logging: %v", err)
+			if err := kubeletConfig.Logging.ValidateAndApplyAsField(utilfeature.DefaultFeatureGate, field.NewPath("logging")); err != nil {
+				klog.ErrorS(err, "Failed to initialize logging")
+				os.Exit(1)
```
why os.Exit instead of returning an error?

pohly replied:
I think that comes from the initial implementation where this code ran in a pre-function which didn't have an error return code. That was changed later. I'll update it.

pohly replied:
Done.

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -237,10 +235,6 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	allErrors = append(allErrors, metrics.ValidateShowHiddenMetricsVersion(kc.ShowHiddenMetricsForVersion)...)
 
-	if errs := logs.ValidateLoggingConfiguration(&kc.Logging, field.NewPath("logging")); len(errs) > 0 {
```
why don't we want to validate logging configuration here any more?

pohly replied:
It's done already earlier in `ValidateAndApply` which all binaries should call as soon as possible to configure logging for the other validate calls - several of those have logging calls. Combining those two actions into one was a conscious decision to ensure that the package gets used as intended.

I was a bit torn whether exposing just `Validate` as API call still makes sense. It might be useful in unit tests. In the end I decided to keep the `logs` API as simple as possible, but that's up for debate.

pohly replied:
I now have a problem for which I do not have a solution yet:
- `pkg/kubelet/apis/config/validation/validation.go` uses `utilfeature.DefaultFeatureGate` when calling `LoggingConfiguration.Validate`
- the new feature gates must be added to  `utilfeature.DefaultMutableFeatureGate` in `validation_test.go`, otherwise looking up the feature panics
- `hack/verify-test-featuregates.sh` [complains about that](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/105797/pull-kubernetes-verify/1520003351708700672):


```
{Script Error ScriptError Test files may not access mutable global feature gates directly:
./pkg/kubelet/apis/config/validation/validation_test.go:79:	logsapi.AddFeatureGates(utilfeature.DefaultMutableFeatureGate)

Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()}
```

Do I have to modify `hack/verify-test-featuregates.sh`, for example to allow usage of `utilfeature.DefaultMutableFeatureGate` in `AddFeatureGates` calls?

liggitt replied:
rather than referring to `utilfeature.DefaultFeatureGate.DeepCopy()` inside `ValidateKubeletConfiguration` and therefore having to mess with it in test setup, I'd probably:
1. pass it as an arg to ValidateKubeletConfiguration
2. pass in `utilfeature.DefaultFeatureGate` from non-test call sites
3. construct a test-only instance in validation_test.go and register log gates into that, e.g.:
   ```go
   testFeatureGates := utilfeature.DefaultFeatureGate.DeepCopy()
   logsapi.AddFeatureGates(testFeatureGates)
   ValidateKubeletConfiguration(..., testFeatureGates)
   ```

pohly replied:
That works, thanks.

pohly commented:
/retest

pohly commented:
Rebased because of a conflict in `staging/src/k8s.io/kubelet/go.sum`. The failure because of utilfeature.DefaultMutableFeatureGate already got resolved earlier.

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -93,8 +93,25 @@ func (c *LoggingConfiguration) Validate(featureGate featuregate.FeatureGate, fld
+	if err != nil {
 		errs = append(errs, field.Invalid(fldPath.Child("format"), c.Format, "Unsupported log format"))
+	} else if factory != nil {
+		feature := factory.Feature()
+		if feature != "" {
```
this seems overly complicated and vulnerable to accidental enablement by default (returning "" effectively makes it stable)

would it be feasible to make formats indicate their stability level when they register, so they *have* to choose a level (alpha/beta/stable)

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -93,8 +93,25 @@ func (c *LoggingConfiguration) Validate(featureGate featuregate.FeatureGate, fld
+				enabled = featureGate.Enabled(feature)
+			}
+			if !enabled {
+				// Look up alpha/beta from actual feature. Should exist, but the API doesn't guarantee it.
+				status := "experimental"
+				if featureSpec, ok := featureGate.DeepCopy().GetAll()[feature]; ok {
+					status = string(featureSpec.PreRelease)
+				}
```
is `feature` ever going to be anything other than `LoggingAlphaOptions` or `LoggingBetaOptions`? I'd recommend just saying "Log format %s is not available because feature %s is not enabled" instead of doing this lookup of the PreRelease field

pohly replied:
Agreed, adding the stability level doesn't add much additional information. Removed.

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -117,7 +134,30 @@ func (c *LoggingConfiguration) Validate(featureGate featuregate.FeatureGate, fld
+	if featureGate != nil {
+		enabled = featureGate.Enabled(LoggingAlphaOptions)
+	}
+	if !enabled {
```
this structure might be correct now, but it's likely that new options will get inserted and will accidentally pick up beta status of other options

I'd suggest structuring this in a way that makes it clear for each option whether it is alpha or beta, something like:

```go
// default to alpha and beta features being disabled
enabled := map[featuregate.Feature]bool{}
if featureGate != nil {
  enabled[LoggingAlphaOptions] = featureGate.Enabled(LoggingAlphaOptions)
  enabled[LoggingBetaOptions] = featureGate.Enabled(LoggingBetaOptions)
}

// make sure alpha and beta config options aren't specified if features aren't enabled

if f := LoggingAlphaOptions; c.Options.JSON.SplitStream && !enabled[f] {
  errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", f)))
}
if f := LoggingAlphaOptions; c.Options.JSON.InfoBufferSize.Value() && !enabled[f] {
  errs = append(errs, field.Forbidden(fldPath.Child("infoBufferSize"), fmt.Sprintf("Feature %s is disabled", f)))
}
```

pohly replied:
I've used a helper function in each if check.

pohly commented on staging/src/k8s.io/component-base/logs/api/OWNERS:
```diff
@@ -0,0 +1,4 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+labels:
+  - kind/api-change
```
Changed, without the sig-instrumentation-api-reviewers alias. If someone wants that, we can add it later.

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/doc.go:
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Well, this was correct when initially written :sweat_smile:  Fixed.

Just to be sure: when moving or modifying existing files, the copyright does not need to be updated?

I would prefer that myself as these copyright lines are informative and not legally binding anyway, and keeping them up-to-date is easy to miss. I just wasn't sure whether Kubernetes has a policy for this.

liggitt replied:
> Just to be sure: when moving or modifying existing files, the copyright does not need to be updated?

that's my understanding

aojea commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -50,58 +61,186 @@ func NewOptions() *Options {
+		c.Format = "text"
+	}
+	if c.FlushFrequency == 0 {
+		c.FlushFrequency = 5 * time.Second
```
you define this as a constant `LogFlushFreqDefault` ? should it be used here?

pohly replied:
Yes, good catch. Fixed.

pohly commented on staging/src/k8s.io/component-base/go.mod:
```diff
@@ -28,6 +28,7 @@ require (
 	k8s.io/client-go v0.0.0
 	k8s.io/klog/v2 v2.60.1
 	k8s.io/utils v0.0.0-20220210201930-3a6ce19ff2f9
+	sigs.k8s.io/yaml v1.2.0
```
Makes sense.

However, sigs.k8s.io/yaml was (and remains) an indirect dependency because of:
```
k8s.io/component-base/codec
k8s.io/apimachinery/pkg/runtime/serializer
k8s.io/apimachinery/pkg/runtime/serializer/json
sigs.k8s.io/yaml
```

pohly committed:
logs: replace config methods with functions

API types are only supposed to have methods related to serialization.

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -50,58 +61,186 @@ func NewOptions() *Options {
 // these features is used.
-func (o *Options) ValidateAndApply(featureGate featuregate.FeatureGate) error {
-	errs := o.validate()
+func (c *LoggingConfiguration) ValidateAndApply(featureGate featuregate.FeatureGate) error {
```
Moving the pflag support into a wrapper type makes sense to me - done and squashed. I kept the helper code exported just in case that it might be needed elsewhere, but right now it is only used internally and thus could also be removed from the API of the package.

I'm less sure about the config functions. It makes the type inconsistent with other, similar types. The godoc documentation becomes harder to read. Callers need an additional import. For example:
```patch
diff --git a/cmd/kube-controller-manager/app/options/options.go b/cmd/kube-controller-manager/app/options/options.go
index 5c7a15b4aac..abe4e4e533b 100644
--- a/cmd/kube-controller-manager/app/options/options.go
+++ b/cmd/kube-controller-manager/app/options/options.go
@@ -35,6 +35,7 @@ import (
        cpoptions "k8s.io/cloud-provider/options"
        cliflag "k8s.io/component-base/cli/flag"
        "k8s.io/component-base/logs"
+       logsapi "k8s.io/component-base/logs/api/v1"
        "k8s.io/component-base/metrics"
        cmoptions "k8s.io/controller-manager/options"
        kubectrlmgrconfigv1alpha1 "k8s.io/kube-controller-manager/config/v1alpha1"
@@ -257,7 +258,7 @@ func (s *KubeControllerManagerOptions) Flags(allControllers []string, disabledBy
        s.SAController.AddFlags(fss.FlagSet("serviceaccount controller"))
        s.TTLAfterFinishedController.AddFlags(fss.FlagSet("ttl-after-finished controller"))
        s.Metrics.AddFlags(fss.FlagSet("metrics"))
-       s.Logs.AddFlags(fss.FlagSet("logs"))
+       logsapi.AddFlags(s.Logs, fss.FlagSet("logs"))
 
        fs := fss.FlagSet("misc")
        fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig).")
```

I've also changed those, but kep that change in a separate commit in case that we want to back it out again.

liggitt replied:
> It makes the type inconsistent with other, similar types

I don't see Validation / Defaulting / AddFlags methods defined on any other API types. The distinct functions lgtm

pohly replied:
I was thinking of types like the ones behind s.Metrics, s.TTLAfterFinishedController, etc. above, like [component-base/metrics.Options](https://github.com/kubernetes/kubernetes/blob/19c8a2127177b43effb9bfe1de272d6f08ea56e8/staging/src/k8s.io/component-base/metrics/options.go#L29-L97). Those are not API types (yet), but perhaps should be.

Anyway, I'm fine with distinct functions, so let's use that.

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/registry.go:
```diff
@@ -41,61 +38,46 @@ type LogFormatFactory interface {
+// RegisterLogFormat registers support for a new logging format. This must be called
+// before using any of the methods in LoggingConfiguration.
+func RegisterLogFormat(name string, factory LogFormatFactory) {
+	logRegistry.register(name, factory)
```
Done.

liggitt replied:
sgtm

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/registry.go:
```diff
@@ -41,61 +38,46 @@ type LogFormatFactory interface {
+func (lfr *logFormatRegistry) register(name string, factory LogFormatFactory) {
 	if lfr.frozen {
-		return fmt.Errorf("log format is frozen, unable to register log format")
+		panic(fmt.Sprintf("log format registry is frozen, unable to register log format %s", name))
```
Done.

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/validate.go:
```diff
@@ -1,70 +0,0 @@
-	"k8s.io/component-base/logs/registry"
-)
-
-func ValidateLoggingConfiguration(c *config.LoggingConfiguration, fldPath *field.Path) field.ErrorList {
```
Done as a separate commit.

pohly commented:
/retest

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/pflags.go:
```diff
@@ -0,0 +1,104 @@
+// String returns the -vmodule parameter (comma-separated list of pattern=N).
+func (wrapper vmoduleConfigurationPFlag) String() string {
+	var patterns []string
+	for _, item := range *wrapper.value {
```
is it possible for value to be nil? if so, do we need a nil guard here?

pohly replied:
No, VModuleConfigurationPflag always must be called with a non-nil value, and that is the only way to construct a wrapper. It wouldn't make sense to call it with nil. Let me add that to the documentation of that function, just to be clear.

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/pflags.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func (wrapper verbosityLevelPflag) String() string {
+	return strconv.FormatInt(int64(*wrapper.value), 10)
```
same question about whether it's possible for value to be nil?

liggitt commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -0,0 +1,282 @@
+
+func validateJSONOptions(c *LoggingConfiguration, featureGate featuregate.FeatureGate, fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	if c.Options.JSON.SplitStream && !featureEnabled(featureGate, LoggingAlphaOptions) {
+		errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", LoggingAlphaOptions)))
```
suggest this format to make sure the lookup and error message use the same level (will prevent accidental drift as we graduate to beta, etc):

```suggestion
	if gate := LoggingAlphaOptions; c.Options.JSON.SplitStream && !featureEnabled(featureGate, gate) {
		errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", gate)))
```

pohly replied:
Good idea, done.

liggitt reviewed (commented):
all the updates look good. a couple questions on the flag wrapper nil cases, a couple nits, and a rebase needed, then this is good to go

thanks for all the work on this, I'm pretty happy with where this ended up

pohly commented on staging/src/k8s.io/component-base/logs/api/v1/options.go:
```diff
@@ -0,0 +1,282 @@
+	"k8s.io/component-base/featuregate"
+)
+
+var (
```
Makes sense. I was thinking that changing that default at the program level might make sense, but that's unnecessary (same effect can be achieved locally by setting the intended default before calling SetRecommendedLoggingConfiguration) and const is cleaner. Changed.

liggitt commented:
lgtm, just needs the rebase. I'd probably leave the distinct commits rather than squashing this one

pohly commented:
Manual changes before rebase: https://github.com/kubernetes/kubernetes/compare/7d7cd6fd4474c6781a1f858bee2077891ff1a0b1..581b94827ad8b467253da1caca23d3f5e4635470

liggitt commented:
/lgtm
/approve

pohly commented:
> I'd probably leave the distinct commits rather than squashing this one

Agreed. The commit messages have the rationale behind each commit, which is easier to correlate with the code when the commits are kept separate.

I've rebased, let's see whether I caught all new code that needed changes (like staging/src/k8s.io/component-base/logs/json/klog_test.go and test/integration/logs/functional/json/output_test.go, which had new references to the old config.VerbosityLevel).

jberkus commented:
As an api chaged, shouldn't this have a release note?

@csantanapr

liggitt commented:
> As an api change, shouldn't this have a release note?

no, there were no externally visible changes... the same config file content still loads successfully.

pohly commented:
There seem to be different opinions about "api change". I was told in https://kubernetes.slack.com/archives/C2C40FMNF/p1650979139995909 that "API change" also includes Go API changes in our staging repos. From that perspective, a release note would be needed here, so I added one.

I don't think that this is done consistently and I keep forgetting about it myself, not the least because I, as a developer who consumes Kubernetes code, don't read the changelog for a release to figure out if my code breaks when I update dependencies. I rather expect obvious compile errors and then figure out by reading the code what I need to change.
